packages/mui-material/src/ComponentWithComponentProp.d.ts

```diff
@@ -0,0 +1,24 @@
+import * as React from 'react';
+import { CommonProps } from './OverridableComponent';
+
+export type Component<Props = any> = React.ElementType<Props>;
+// type Example = Component<{ prop: string }>
+
+export type PropsWithComponentProp<Props = {}, Type extends Component = Component> = Props &
+  CommonProps &
```

Comment:
I suppose we'll need another version of this type for non-material components (e.g. Base). `CommonProps` extends `StyledComponentsProps` that does not apply to Base components.

Reply:
Agree, I would like to keep this as close to the `OverridableComponent` as possible. We could have something similar to this for the base components, something like: `BasePropsWithComponentProp`.

Reply:
I've just realized we've got two `OverridableComponent` definitions right now - one lives in @mui/material and has props with `StyledComponentsProps` and the other is in @mui/core.
I agree they should be named differently to avoid mistakes.